adserver/analyzer/backends/textacynlp.py

```diff
@@ -0,0 +1,95 @@
+"""Textacy/Spacy-based keyword analyzer that lemmatizes and uses textrank for keyphrase extraction."""
+import collections
+
+import textacy
+from textacy import preprocessing
+
+from .naive import NaiveKeywordAnalyzerBackend
+
+
+class TextacyAnalyzerBackend(NaiveKeywordAnalyzerBackend):
+
+    """
+    An analyzer backend that uses textacy for lemmatization and keyword extraction.
+
+    https://textacy.readthedocs.io/en/latest/quickstart.html
+    """
+
+    TOP_PHRASE_COUNT = 20
+
+    # Minimum phrase length where each word isn't required to be in the output phrase
+    MIN_PHRASE_LENGTH = 6
+
+    def __init__(self, url, **kwargs):
+        """Overrides to lemmatize keywords."""
+        super().__init__(url, **kwargs)
+
+        self.pretrained_model = textacy.load_spacy_lang(
+            "en_core_web_md", disable=("parser",)
+        )
+
+        self.preprocessor = preprocessing.make_pipeline(
+            preprocessing.normalize.unicode,
+            preprocessing.remove.punctuation,
+            preprocessing.normalize.whitespace,
+        )
+
+        # Maps lemmatized keywords back to the original keyword/phrase in our database
+        self.lemmatized_to_original_map = {}
+
+        # Reset the keywords
+        self.keyword_corpus = []
+
+        # Build up the lemmatized keywords
+        for topic in self.topics:
+            for keyword in self.topics[topic]:
+                terms_doc = textacy.make_spacy_doc(
+                    self.preprocessor(keyword).lower(), lang=self.pretrained_model
+                )
+                words = list(
+                    textacy.extract.utils.terms_to_strings(terms_doc, by="lemma")
+                )
+                phrase = " ".join(words)
+                self.keyword_corpus.append(phrase)
+                self.lemmatized_to_original_map[phrase] = keyword
+
+    def preprocess_text(self, text):
+        return self.preprocessor(text).lower()
+
+    def analyze_text(self, text):
+        """
+        Analyze a large string of text for keyword extraction.
+
+        One notable difference from the other backends is that the matching
+        keywords/phrases should be in order of how used/present they are.
+        """
+        doc = textacy.make_spacy_doc(text, lang=self.pretrained_model)
+        keyword_counter = collections.Counter()
+
+        # Lemmatize the words/phrases to make sure we're always finding exactly what we're looking for
+        # Bias toward the early part of the document
+        for phrase, weight in textacy.extract.keyterms.textrank(
+            doc, normalize="lemma", position_bias=True, topn=self.TOP_PHRASE_COUNT
```

Comment:
Interesting that we aren't inputting our keywords here. It seems like that would help make more educated guesses around what we're wanting to pull out. Seems like there's definitely some interesting optimization possibilities here with other models, etc.

Reply:
I haven't found something exactly like that in any library I looked at. Instead of analyzing some text and trying to reason about it, you're trying to classify how close a text is to some predefined topic set. I didn't find something that did that exactly.